---
title: "sbsw_mt_revisions_binwise"
output: html_document
date: "2024-02-06"
---

This script contains revised data analysis and figures for the SBSW manuscript while retaining the bin-wise nature of the differential expression analysis (ie each DESeq2 analysis was generated by the genes specific to a given bin, then all stitched together into one results table)

Pros: 
1. Specificity of understanding up/down-regulated genes in a given genome, genome-wise trends

Cons:
1. Are p-values and LFC values comparable if they are not generated all together? 

#load packages 
```{r}
library(tidyverse)
library(DESeq2)

```

#load data
```{r}
#load data
mt_tally <-read_csv("../current_data/mt_tally_v2.csv") #raw data from bwa
metadata_norm_v2 <-read_csv("metadata_norm_v2.csv") #metadata
master_tax <-read_csv("../current_data/master_genes_taxonomy_gtdbtk_blocked.csv") %>% select(-"...1")
diffex_tax <-master_tax %>% filter(padj <0.05)
normalized_counts <-read_csv("../current_data/normalized_mt_tally_v2_genes.csv")
corrected_counts <-read_csv("../current_data/limma_corrected_counts.csv")

#load visuals
plot_theme <- theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.5), axis.text = element_text(color = "black"))
```

#MA plot of fold change vs base mean for all genes in the dataset

The DESeq2 internal MA plot uses the function plotMA from geneplotter. Since we no longer have a deseq2 dataset format, I can simply use geneplotter to generate my MA plot.
```{r}
install.packages("geneplotter")
library(geneplotter)

master_tax <-as.data.frame(master_tax)
master_tax$significance <-"FALSE"
master_tax$significance[master_tax$padj < 0.05] <- "TRUE"
master_tax$significance <-as.logical(master_tax$significance)

MA_plot <-master_tax %>% select(-c(3:14))
MA_plot <-MA_plot[,c("baseMean", "log2FoldChange", "significance")]
plotMA(MA_plot, ylim=c(-3,3))

```

#PCA plot of overall sample differences 

PCA test to visualize batch effects on the data in DESeq:
```{r}

#create matrix from mt_tally and call this mt_raw
mt_raw <- mt_tally[,c(1, 5:40)]
mt_raw <- as.data.frame(mt_raw)
rownames(mt_raw) <-mt_raw$position
mt_raw[,"position"] <- NULL

dds <- DESeqDataSetFromMatrix(countData=mt_raw,
                              colData=metadata_norm_v2,
                              design=~extraction_batch + timecount_norm) #load deseq2 dataset 


dds <- estimateSizeFactors( dds ) #normalizing data by gene length and rna counts

save(dds, file="raw_deseq_dataset.Rdata")

sizeFactors(dds)

vsd <-vst(dds, blind=FALSE) #variance stabilizing transformation that is not blind to design. Design is used to calculate variability but not remove it 

plotPCA(vsd, intgroup=c("time24", "extraction_batch")) + scale_color_brewer(palette="Paired") #visualize the PCA plot of data before batch effects are removed-- batch accounts for ~22% of the variation and is a significant variable

```

Using limma to remove batch effects from the variance stabilized count matrix:

```{r}
mat <- assay(vsd)
mm <- model.matrix(~timecount_norm, colData(vsd))
mat <- limma::removeBatchEffect(mat, batch=vsd$extraction_batch, design=mm)
assay(vsd) <- mat
plotPCA(vsd, intgroup=c("time24", "extraction_batch", "pH", "temperature", "DO_sat", "Aw (in situ)", "light_condition")) + scale_color_brewer(palette="Paired")

#this limma-corrected matrix can be used for downstream applications independent of deseq2 such as pca, statistical testing, etc. it should NOT be used as the input for deseq2 as it is not raw count data! deseq2 will remove batch effects internally if they are included in the design 


write.csv(mat, "limma_corrected_counts.csv")

#PCA plot outside deseq2
pcadata <- plotPCA(vsd, intgroup=c("time24", "extraction_batch", "pH", "temperature", "DO_sat", "Aw (in situ)", "light_condition"), returnData=TRUE)
#This gives me the position of each sample within PC1 and PC2, but not any other information about PCs. Good for plotting but less good for analysis
```

How best to visualize the whole dataset?  showing overall differences btwn samples and that there is truly a day/night difference? 

PCA plot using prcomp and ggbiplot:
https://tavareshugo.github.io/data-carpentry-rnaseq/03_rnaseq_pca.html
```{r}
library(devtools)
library(ggbiplot)

#Put dataframe of limma-corrected counts in a format accepted by prcomp

#pcadf <-corrected_counts %>% pivot_longer(!position, names_to="sample", values_to="corrected_count") %>% left_join(., metadata_norm_v2[,c(2, 5:10, 12)], by="sample")

#create data matrix from corrected_counts

pca_matrix <- corrected_counts %>% column_to_rownames("position") %>% as.matrix() %>% t()
#see how it looks 
pca_matrix[1:10, 1:5] #samples are rows, columns are variables (genes)

sample_pca <-prcomp(pca_matrix)
pc_eigenvalues <-sample_pca$sdev^2

# create a "tibble" manually with 
# a variable indicating the PC number
# and a variable with the variances
pc_eigenvalues <- tibble(PC = factor(1:length(pc_eigenvalues)), 
                         variance = pc_eigenvalues) %>% 
  # add a new column with the percent variance
  mutate(pct = variance/sum(variance)*100) %>% 
  # add another column with the cumulative variance explained
  mutate(pct_cum = cumsum(pct))

# print the result
pc_eigenvalues

#create scree plot to viaualize this
pc_eigenvalues %>% 
  ggplot(aes(x = PC)) +
  geom_col(aes(y = pct)) +
  geom_line(aes(y = pct_cum, group = 1)) + 
  geom_point(aes(y = pct_cum)) +
  labs(x = "Principal component", y = "Fraction variance explained")
```



#Looking for most highly expressed genomes

```{r}
hits_per_bin <- master_tax %>% group_by(bin) %>% summarize(hits_per_bin=length(bin))

diffex_hits <-diffex_tax %>% group_by(bin) %>% summarize(diffex_hits=length(bin)) 

prop_diffex <-left_join(diffex_hits, hits_per_bin, by="bin") %>% mutate(prop_genome_diffex=diffex_hits/hits_per_bin)
```

Many of these genomes are the same as those I've identified with interesting results in sod, etc

#bacteriorhodopsins
How to locate real bacteriorhodopsin genes vs homologs? 


```{r}
rhodopsin <-diffex_tax %>% filter(product=="bacteriorhodopsin")

all_rhodopsin <-master_tax %>% filter(product=="bacteriorhodopsin")

#creating a line plot showing the relative expression of these genes to their base mean value
#this cannot be done for batch-corrected count data since it is log transformed

#counts for all bacteriorhodopsin genes (not corrected!)
rhodopsin_counts <-normalized_counts %>% filter(position %in% rhodopsin$position) %>% pivot_longer(c(1:36), names_to="sample") %>% left_join(., metadata_norm_v2, by="sample") %>% group_by(time24, position) 

#relative expression (divide by basemean)
rhodopsin_relative <-rhodopsin_counts %>% left_join(rhodopsin[,c(2,4)], by="position") %>% mutate(relative_expression=value/baseMean)
rhodrelsum <-rhodopsin_relative %>% group_by(time24, position) %>% summarize(relative_expression_mean=mean(relative_expression), sd=sd(relative_expression)) %>% left_join(., rhodopsin[,c(4:7,12)], by="position")


#line plot with error bars
ggplot(rhodrelsum, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd, fill="grey")) +
  geom_line(aes(color=family)) +
  geom_point(aes(color=family)) +
  scale_color_manual(values=c("blue", "green")) +
  plot_theme



```

Visualizing salinibacter ruber expression in SBSW:

201005_12_DNA_2_bin.30, 201005_12_DNA_mat_2_bin.6, 201005_12_DNA_2_bin.48

```{r}

bin48 <-diffex_tax %>% filter(bin=="201005_12_DNA_2_bin.48")
```







Thoughts on visualizing these genes
-Could create a basemean cutoff to exclude low-count genes. This would have to be consistent across the entire dataset. 

