---
title: "sbsw_mt_revisions_binwise"
output: html_document
date: "2024-02-06"
---

This script contains revised data analysis and figures for the SBSW manuscript while retaining the bin-wise nature of the differential expression analysis (ie each DESeq2 analysis was generated by the genes specific to a given bin, then all stitched together into one results table)

Pros: 
1. Specificity of understanding up/down-regulated genes in a given genome, genome-wise trends

Cons:
1. Are p-values and LFC values comparable if they are not generated all together? 

#load packages 
```{r}
library(tidyverse)
library(DESeq2)
library(wesanderson)

```

#load data
```{r}
#load data
mt_tally <-read_csv("../current_data/mt_tally_v2.csv") #raw data from bwa
metadata_norm_v2 <-read_csv("metadata_norm_v2.csv") #metadata
master_tax <-read_csv("../current_data/master_genes_taxonomy_gtdbtk_blocked.csv") %>% select(-"...1")
diffex_tax <-master_tax %>% filter(padj <0.05)
normalized_counts <-read_csv("../current_data/normalized_mt_tally_v2_genes.csv")
corrected_counts <-read_csv("../current_data/limma_corrected_counts.csv")

#load visuals
plot_theme <- theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5), axis.text = element_text(color = "black"))

wes_palette_27 <-c(wes_palette("Cavalcanti1"), wes_palette("GrandBudapest1"), wes_palette("Moonrise3"), wes_palette("GrandBudapest2"), wes_palette("IsleofDogs1"), wes_palette("Darjeeling1"))


```

#MA plot of fold change vs base mean for all genes in the dataset

The DESeq2 internal MA plot uses the function plotMA from geneplotter. Since we no longer have a deseq2 dataset format, I can simply use geneplotter to generate my MA plot.
```{r}
install.packages("geneplotter")
library(geneplotter)

master_tax <-as.data.frame(master_tax)
master_tax$significance <-"FALSE"
master_tax$significance[master_tax$padj < 0.05] <- "TRUE"
master_tax$significance <-as.logical(master_tax$significance)

MA_plot <-master_tax %>% select(-c(3:14))
MA_plot <-MA_plot[,c("baseMean", "log2FoldChange", "significance")]
plotMA(MA_plot, ylim=c(-3,3))

#Alternatively, volcano plot

master_tax$diffex <-"padj > 0.05"
master_tax$diffex[master_tax$padj <0.05] <-"padj <0.05"
master_tax$diffex[master_tax$padj < 0.01] <- "padj < 0.01"

ggplot(master_tax, aes(x=log2FoldChange, y=-log10(padj), col=diffex)) + geom_point() +
  scale_color_manual(values=c("blue", "red", "black"))



```

#PCA plot of overall sample differences 

PCA test to visualize batch effects on the data in DESeq:
```{r}

#create matrix from mt_tally and call this mt_raw
mt_raw <- mt_tally[,c(1, 5:40)]
mt_raw <- as.data.frame(mt_raw)
rownames(mt_raw) <-mt_raw$position
mt_raw[,"position"] <- NULL

dds <- DESeqDataSetFromMatrix(countData=mt_raw,
                              colData=metadata_norm_v2,
                              design=~extraction_batch + timecount_norm) #load deseq2 dataset 


dds <- estimateSizeFactors( dds ) #normalizing data by gene length and rna counts

save(dds, file="raw_deseq_dataset.Rdata")
load("../current_data/raw_deseq_dataset.Rdata")

sizeFactors(dds)

vsd <-vst(dds, blind=FALSE) #variance stabilizing transformation that is not blind to design. Design is used to calculate variability but not remove it 

plotPCA(vsd, intgroup=c("time24", "extraction_batch")) + scale_color_brewer(palette="Paired") #visualize the PCA plot of data before batch effects are removed-- batch accounts for ~22% of the variation and is a significant variable

```

Using limma to remove batch effects from the variance stabilized count matrix:

```{r}
mat <- assay(vsd)
mm <- model.matrix(~timecount_norm, colData(vsd))
mat <- limma::removeBatchEffect(mat, batch=vsd$extraction_batch, design=mm)
assay(vsd) <- mat
plotPCA(vsd, intgroup=c("time24", "extraction_batch_numeric", "pH", "temperature", "DO_sat", "Aw (in situ)")) + scale_color_brewer(palette="Paired")

#this limma-corrected matrix can be used for downstream applications independent of deseq2 such as pca, statistical testing, etc. it should NOT be used as the input for deseq2 as it is not raw count data! deseq2 will remove batch effects internally if they are included in the design 


write.csv(mat, "limma_corrected_counts.csv")

#PCA plot outside deseq2
pcadata <- plotPCA(vsd, intgroup=c("time24", "extraction_batch", "pH", "temperature", "DO_sat", "Aw (in situ)", "light_condition", "time"), returnData=TRUE)
#This gives me the position of each sample within PC1 and PC2, but not any other information about PCs. Good for plotting but less good for analysis
#it uses the assay(vsd) which is the limma transformed dataset
#uses the top 500 genes by row variance to determine PCs

plotPCA(vsd, intgroup=c("time24", "extraction_batch",  "light_condition"), ntop=1000) +
  scale_color_brewer(palette="Paired") #variance explained by PCs decreases as more genes are used in the assay


```

See the following explanation for why only the top 500 genes with greatest variance across samples were used to create PCA:
https://support.bioconductor.org/p/51270/
"For (i)[sample clustering via ordination], it does not make much difference whether you use all data or only highly variable genes, as genes with low variance across samples
provide only little information on sample distances anyway and so have
little influence on the result.

For (ii)[gene clustering], it is common practice to subset to the most highly variant
genes because *many ordination or clustering methods do not cope well
with a matrix with thousands of rows, and the genes with low variance
are unlikely to be part of interesting clusters anyway*."

#Plotting PCA in ggplot to customize graphics:

```{r}
pcadata <- plotPCA(vsd, intgroup=c("time24", "extraction_batch", "pH", "temperature", "DO_sat", "Aw (in situ)", "light_condition", "time"), returnData=TRUE)
percentVar <- round(100 * attr(pcadata, "percentVar"))

ggplot(pcadata, aes(x=PC1, y=PC2)) +
  geom_point(size=3, aes(color=time, shape=extraction_batch)) +
  stat_ellipse(mapping=aes(linetype=light_condition)) +
  scale_color_brewer(palette="Paired") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  plot_theme
  
```

#sample heatmap

```{r}
library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("extraction_batch","time", "light_condition")])
pheatmap(assay(vsd)[select,], cluster_rows=TRUE, show_rownames=FALSE,
         cluster_cols=TRUE, annotation_col=df)
```


How best to visualize the whole dataset?  showing overall differences btwn samples and that there is truly a day/night difference? 

"Note that lack of clustering in a PCA does not mean nothing is happening. Imagine an RNAseq experiment where only a dozen genes changed due to your treatment. Because thereâ€™s few genes, transcriptome-wide you might not capture structure in the data."
PCA plot using prcomp and ggbiplot:
https://tavareshugo.github.io/data-carpentry-rnaseq/03_rnaseq_pca.html
```{r}
library(devtools)
library(ggbiplot)

#Put dataframe of limma-corrected counts in a format accepted by prcomp

#pcadf <-corrected_counts %>% pivot_longer(!position, names_to="sample", values_to="corrected_count") %>% left_join(., metadata_norm_v2[,c(2, 5:10, 12)], by="sample")

#create data matrix from corrected_counts

pca_matrix <- corrected_counts %>% column_to_rownames("position") %>% as.matrix() %>% t()
#see how it looks 
pca_matrix[1:10, 1:5] #samples are rows, columns are variables (genes)

sample_pca <-prcomp(pca_matrix)
pc_eigenvalues <-sample_pca$sdev^2

# create a "tibble" manually with 
# a variable indicating the PC number
# and a variable with the variances
pc_eigenvalues <- tibble(PC = factor(1:length(pc_eigenvalues)), 
                         variance = pc_eigenvalues) %>% 
  # add a new column with the percent variance
  mutate(pct = variance/sum(variance)*100) %>% 
  # add another column with the cumulative variance explained
  mutate(pct_cum = cumsum(pct))

# print the result
pc_eigenvalues

#create scree plot to viaualize this
pc_eigenvalues %>% 
  ggplot(aes(x = PC)) +
  geom_col(aes(y = pct)) +
  geom_line(aes(y = pct_cum, group = 1)) + 
  geom_point(aes(y = pct_cum)) +
  labs(x = "Principal component", y = "Fraction variance explained")

pc_scores <- sample_pca$x
pc_scores <- pc_scores %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "sample")

# print the result
pc_scores

pc_scores %>% 
  # create the plot
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point()
```

PC1 and PC2 do NOT explain most of the variation in the dataset. Perhaps a non-linear method, like NMDS, would work better? Unsure why it looks so different from the PCA plot generated with plotPCA in DESeq2?
#Could try it on un-corrected data?
```{r}
pca_matrix <- normalized_counts[,c(1:37)] %>% column_to_rownames("position") %>% as.matrix() %>% t() 
#see how it looks 
pca_matrix[1:10, 1:5] #samples are rows, columns are variables (genes)

sample_pca <-prcomp(pca_matrix, scale=TRUE)
pc_eigenvalues <-sample_pca$sdev^2

# create a "tibble" manually with 
# a variable indicating the PC number
# and a variable with the variances
pc_eigenvalues <- tibble(PC = factor(1:length(pc_eigenvalues)), 
                         variance = pc_eigenvalues) %>% 
  # add a new column with the percent variance
  mutate(pct = variance/sum(variance)*100) %>% 
  # add another column with the cumulative variance explained
  mutate(pct_cum = cumsum(pct))

# print the result
pc_eigenvalues

#create scree plot to viaualize this
pc_eigenvalues %>% 
  ggplot(aes(x = PC)) +
  geom_col(aes(y = pct)) +
  geom_line(aes(y = pct_cum, group = 1)) + 
  geom_point(aes(y = pct_cum)) +
  labs(x = "Principal component", y = "Fraction variance explained")

pc_scores <- sample_pca$x
pc_scores <- pc_scores %>% 
  # convert to a tibble retaining the sample names as a new column
  as_tibble(rownames = "sample")

# print the result
pc_scores

pc_scores %>% 
  # create the plot
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point()

#scaled or unscaled, it's a mess
```

#How about NMDS?
#Nope it looks confusing 
```{r}
library(vegan)
pca_matrix <- corrected_counts %>% column_to_rownames("position") %>% as.matrix() %>% t() 
#see how it looks 
pca_matrix[1:10, 1:5] #samples are rows, columns are variables (genes)

#stress test 
dim1<-metaMDS(pca_matrix,k=1)
dim1
s1<-  0.1623909   
s1<-dim1[[22]]
s1
dim2<-metaMDS(pca_matrix,k=2)
s2<-dim2[[22]]
s2
dim3<-metaMDS(pca_matrix,k=3)
s3<-dim3[[22]]
dim4<-metaMDS(pca_matrix,k=4)
s4<-dim4[[22]]
dim5<-metaMDS(pca_matrix,k=5)
s5<-dim5[[22]]
stress.plot<-c(s1,s2,s3,s4,s5)
dim.plot<-c(1,2,3,4,5)
stress.plot
plot(dim.plot,stress.plot, main="STRESS PLOT", pch=19, col="red")

ord <- metaMDS(pca_matrix, k=3)
data.scores=as.data.frame(scores(ord))
data.scores$sample <-rownames(data.scores)

nmds_plot <-left_join(data.scores, metadata_norm_v2, by="sample")

#color is time
ggplot(nmds_plot, aes(x=NMDS1, y=NMDS2)) +
  geom_point(size=3, aes(color=time, shape=extraction_batch)) + 
  plot_theme 
  
#the nmds axes are really tiny suggesting this is not the best way to visualize differences in log-normalized data

```


#Looking for most highly expressed genomes

```{r}
genes_per_bin <- master_tax %>% group_by(bin) %>% summarize(genes_per_bin=length(bin))

diffex_genes <-diffex_tax %>% group_by(bin) %>% summarize(diffex_genes=length(bin)) 

prop_diffex <-left_join(diffex_hits, hits_per_bin, by="bin") %>% mutate(prop_genome_diffex=diffex_hits/hits_per_bin)
```

Many of these genomes are the same as those I've identified with interesting results in sod, etc

#bacteriorhodopsins
How to locate real bacteriorhodopsin genes vs homologs? 
```{r}
rhodopsin <-diffex_tax %>% filter(product=="bacteriorhodopsin")
#n=24 genes
ggplot(rhodopsin, aes(x=position, y=baseMean, fill=family)) +
  geom_bar(stat="identity")

all_rhodopsin <-master_tax %>% filter(product=="bacteriorhodopsin")

#creating a line plot showing the relative expression of these genes to their base mean value
#this cannot be done for batch-corrected count data since it is log transformed

#counts for all bacteriorhodopsin genes (not corrected!)

rhodopsin <- rhodopsin %>% mutate(regulation=if_else(log2FoldChange>0, "up", "down"))
rhodopsin_counts <-normalized_counts %>% filter(position %in% rhodopsin$position) %>% pivot_longer(c(1:36), names_to="sample") %>% left_join(., metadata_norm_v2, by="sample") %>% group_by(time24, position) 

#relative expression (divide by basemean)
rhodopsin_relative <-rhodopsin_counts %>% left_join(rhodopsin[,c(2,4)], by="position") %>% mutate(relative_expression=value/baseMean)
rhodrelsum <-rhodopsin_relative %>% group_by(time24, position) %>% summarize(relative_expression_mean=mean(relative_expression), sd=sd(relative_expression)) %>% left_join(., rhodopsin[,c(4:7,12,15)], by="position")


#line plot with error bars
ggplot(rhodrelsum, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="grey") +
  geom_line(aes(linetype=regulation)) +
  geom_point() +
  scale_linetype_manual(values=c("dotdash","solid")) +
  plot_theme


neg_foldchange <- diffex_tax %>% filter(product=="bacteriorhodopsin" & log2FoldChange <0)
#counts for all bacteriorhodopsin genes (not corrected!)
neg_counts <-normalized_counts %>% filter(position %in% neg_foldchange$position) %>% pivot_longer(c(1:36), names_to="sample") %>% left_join(., metadata_norm_v2, by="sample") %>% group_by(time24, position) 

#relative expression (divide by basemean)
neg_relative <-neg_counts %>% left_join(neg_foldchange[,c(2,4)], by="position") %>% mutate(relative_expression=value/baseMean)
negsum <-neg_relative %>% group_by(time24, position) %>% summarize(relative_expression_mean=mean(relative_expression), sd=sd(relative_expression)) %>% left_join(., neg_foldchange[,c(4:7,12)], by="position")


#line plot with error bars
ggplot(negsum, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="grey") +
  geom_line() +
  geom_point() +
  plot_theme


```

#sod

```{r}
sod <-master_tax %>% filter(., product== "superoxide dismutase") %>% filter(padj <=0.05) #n=15 genes
mean(sod$baseMean) #okay to take the mean of a mean?
1224.632
sodcounts <-normalized_counts %>% filter(position %in% sod$position) %>% pivot_longer(c(1:36), names_to="sample") %>% left_join(., metadata_norm_v2, by="sample")

sodcounts <- sodcounts %>% left_join(sod[,c(2,4)], by="position") %>% mutate(relative=value/baseMean)
sodrelsummary <- sodcounts %>% group_by(time24, position) %>% summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% left_join(., sod[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(sodrelsummary, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="grey") +
  geom_line() +
  geom_point() +
  plot_theme
```

#glycerol kinase

```{r}
glycerol <- diffex_tax[str_detect(diffex_tax$product, "glycerol kinase"),] #n=29 genes, 23 annotated specifically as glpK
#mean expression
mean(glycerol$baseMean)
898.7927
glycerol$gene <-"glycerol kinase"

glycerol_plot <-normalized_counts %>% filter(position %in% glycerol$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(glycerol[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., glycerol[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(glycerol_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```

#aconitate hydratase
```{r}
aco <-diffex_tax[str_detect(diffex_tax$product, "aconitate hydratase"),] #n=11

aco_plot <-normalized_counts %>% filter(position %in% aco$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(aco[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., aco[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(aco_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme
```


#pufMHL
Note that this has a very low base mean of expression
```{r}
photo <- diffex_tax[str_detect(diffex_tax$product, "photosynthetic reaction center"),] #n=11, note that these have very low base mean expression 

photo_plot <-normalized_counts %>% filter(position %in% photo$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(photo[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., photo[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(photo_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme


```

#bacteriochlorophyll
Note that this has a very low base mean of expression
ATP-dependent prochlorophyllide reductase is non-light-dependent, oxygen-sensitive, found in Halothece
chlorophyllide a reductase is key enzyme for producing bacteriochlorophyll
```{r}
chlorophyll <- diffex_tax[str_detect(diffex_tax$product, "chloro"),] 

bchl_plot <-normalized_counts %>% filter(position %in% chlorophyll$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(chlorophyll[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., chlorophyll[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(bchl_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```

#oxygenic photosynthesis
```{r}
photosystem2 <- diffex_tax[str_detect(diffex_tax$product, "photosystem II"),]

photosystem2_plot <-normalized_counts %>% filter(position %in% photosystem2$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(photosystem2[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., photosystem2[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(photosystem2_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

photosystemI <-diffex_tax[str_detect(diffex_tax$product, "photosystem I core protein"),]
photosystemI_plot <-normalized_counts %>% filter(position %in% photosystemI$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(photosystemI[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., photosystemI[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(photosystemI_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```

#Fe-S cluster assembly
```{r}
suf <- diffex_tax[str_detect(diffex_tax$product, "Suf"),] 

suf_plot <-normalized_counts %>% filter(position %in% suf$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(suf[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., suf[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(suf_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```

#Phosphate uptake
Pst only not PhoU
```{r}
phosphate <- diffex_tax %>% filter(product== "phosphate ABC transporter permease PstA"| product=="phosphate ABC transporter ATP-binding protein PstB" | product=="phosphate ABC transporter permease subunit PstC" | product=="PstS family phosphate ABC transporter substrate-binding protein")

phosphate_plot <-normalized_counts %>% filter(position %in% phosphate$position) %>%
pivot_longer(c(1:36), names_to="sample") %>% 
  left_join(phosphate[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., phosphate[,c(4:6, 9,12,13)], by="position")

ggplot(phosphate_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```

ADDING SIX MORE GENES to make it 4x4 not 3x3 figure:

#isocitrate dehydrogenase

```{r}
iso <-diffex_tax[str_detect(diffex_tax$product, "isocitrate dehydrogenase"),]

iso_plot <-normalized_counts %>% filter(position %in% iso$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(iso[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., iso[,c(4:6, 9,12,13)], by="position")

#line plot with error bars
ggplot(iso_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```


#phosphoenolpyruvate synthase

```{r}
pyruv <-diffex_tax %>% filter(product=="phosphoenolpyruvate synthase")

pyruv_plot <-normalized_counts %>% filter(position %in% pyruv$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(pyruv[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., pyruv[,c(4:6, 9,12,13)], by="position")

ggplot(pyruv_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

#No clear pattern--it is used in both glycolysis and gluconeogenesis!
```

#2-isopropylmalate synthase
```{r}

malate <-diffex_tax %>% filter(product=="2-isopropylmalate synthase")

malate_plot <-normalized_counts %>% filter(position %in% malate$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(malate[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., malate[,c(4:6, 9,12,13)], by="position")

ggplot(malate_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme
```

#acetolactate synthase

```{r}
aceto <-diffex_tax[str_detect(diffex_tax$product, "acetolactate synthase"),]

aceto_plot <-normalized_counts %>% filter(position %in% aceto$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(aceto[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., aceto[,c(4:6, 9,12,13)], by="position")

ggplot(aceto_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme
```

#glutamate synthase

```{r}
#glutamate synthase 
glutamate <- diffex_tax %>% filter(., product=="glutamate synthase large subunit")

gluta_plot <-normalized_counts %>% filter(position %in% glutamate$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(glutamate[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., glutamate[,c(4:6, 9,12,13)], by="position")

ggplot(gluta_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

```

#phosphate uptake regulator PhoU

```{r}
phosreg <- diffex_tax %>% filter(product=="phosphate uptake regulator PhoU" | product== "phosphate signaling complex protein PhoU" )

phosreg_plot <-normalized_counts %>% filter(position %in% phosreg$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(phosreg[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., phosreg[,c(4:6, 9,12,13)], by="position")

ggplot(phosreg_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme
```

#bacteriochlorophyll

```{r}
chlorophyll <- diffex_tax[str_detect(diffex_tax$product, "chloro"),] 

chloro_plot <-normalized_counts %>% filter(position %in% chlorophyll$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(chlorophyll[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., chlorophyll[,c(4:6, 9,12,13)], by="position")

ggplot(chloro_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme
```

#glycerol dehydrogenases (for supplement)
```{r}
dehydro <-diffex_tax[str_detect(diffex_tax$product, "glycerol-3-phosphate dehydrogenase"), ]
#make gene names wrappable
dehydro <- dehydro %>% mutate(gene=case_when(
  startsWith(product, "anaerobic") ~ "anaerobic glycerol-3-phosphate dehydrogenase",
  startsWith(product, "glycerol-3-phosphate") ~ "glycerol-3-phosphate dehydrogenase"))

dehydro_plot <-normalized_counts %>% filter(position %in% dehydro$position) %>% 
  pivot_longer(c(1:36), names_to="sample")%>% 
  left_join(dehydro[,c(2,4)], by="position") %>% 
  mutate(relative=value/baseMean) %>% 
  left_join(., metadata_norm_v2[,c(2,4)], by="sample") %>% 
  group_by(time24, position) %>% 
  summarise(relative_expression_mean=mean(relative), sd=sd(relative)) %>% 
  left_join(., dehydro[,c(4:7, 9,12,13)], by="position")

ggplot(dehydro_plot, aes(x=time24, y=relative_expression_mean, group=position)) +
  geom_ribbon(aes(ymin=relative_expression_mean-sd, ymax=relative_expression_mean+sd), fill="gray") +
  geom_line(aes(linetype=gene)) +
  geom_point() +
  plot_theme
```


Visualizing salinibacter ruber expression in SBSW:

201005_12_DNA_2_bin.30, 201005_12_DNA_mat_2_bin.6, 201005_12_DNA_2_bin.48

```{r}

bin48 <-master_tax %>% filter(bin=="201005_12_DNA_2_bin.48")
bin30 <-master_tax %>% filter(bin=="201005_12_DNA_2_bin.30")
bin6 <-master_tax %>% filter(bin=="201005_12_DNA_mat_2_bin.6")
```

MA plots for 3 S. ruber bins as well as the S. ruber genome, vertically aligned 


```{r}

#bin48
bin48 <-as.data.frame(bin48)
bin48$significance <-"FALSE"
bin48$significance[bin48$padj < 0.05] <- "TRUE"
bin48$significance <-as.logical(bin48$significance)

bin48_diffex <- bin48 %>% filter(padj<0.05)
MA_plot <-bin48[,c("baseMean", "log2FoldChange", "significance")]
geneplotter::plotMA(MA_plot, ylim=c(-2,2), colSig="blue", colLine="#ff000080")

#bin30
bin30 <-as.data.frame(bin30)
bin30$significance <-"FALSE"
bin30$significance[bin30$padj < 0.05] <- "TRUE"
bin30$significance <-as.logical(bin30$significance)

bin30_MA_plot <-bin30[,c("baseMean", "log2FoldChange", "significance")]
geneplotter::plotMA(bin30_MA_plot, ylim=c(-2,2), colSig="blue")

#bin6
bin6 <-as.data.frame(bin6)
bin6$significance <-"FALSE"
bin6$significance[bin6$padj < 0.05] <- "TRUE"
bin6$significance <-as.logical(bin6$significance)

bin6_MA_plot <-bin6[,c("baseMean", "log2FoldChange", "significance")]
geneplotter::plotMA(bin6_MA_plot, ylim=c(-2,2), colSig="red")

```

#investigating rhodopsins in salinibacter bins and culture

```{r}

salinibacter <- master_tax %>% filter(bin %in% c("201005_12_DNA_2_bin.48", "201005_12_DNA_2_bin.30", "201005_12_DNA_mat_2_bin.6"))
salinibacter_diffex <- salinibacter %>% filter(padj<=0.05)

#load s ruber culture files

load("sruber_light_deseq.Rdata")
culture_diffex <-light_df %>% filter(padj<=0.05)
load("sruber_normalized_counts.Rdata")
metadata <-read_csv("../sruber/sruber_cds_alignment/sruber_metadata.csv")

#what are the xanthorhodopsins doing? NC_007677.1_WP_011404249.1, NC_007677.1_WP_011405291.1, NC_007677.1_WP_011405484.1, NC_007677.1_WP_011405223.1

xantho_plot <-normalized_counts %>% filter(gene_id %in% c("NC_007677.1_WP_011404249.1", "NC_007677.1_WP_011405291.1", "NC_007677.1_WP_011405484.1", "NC_007677.1_WP_011405223.1")) %>% pivot_longer(!gene_id, names_to="sample") %>% left_join(metadata, by="sample") %>% group_by(gene_id, date_time) %>% summarise(mean_exp=mean(value), sd=sd(value))

#line plot with error bars
ggplot(xantho_plot, aes(x=date_time, y=mean_exp, group=gene_id)) +
  geom_ribbon(aes(ymin=mean_exp-sd, ymax=mean_exp+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme

#one xanthorhodopsin only turned on in the light! as expected. phew
#all the others have much less expression

ggplot(xantho_plot %>% filter(!(gene_id=="NC_007677.1_WP_011404249.1")), aes(x=date_time, y=mean_exp, group=gene_id)) +
  geom_ribbon(aes(ymin=mean_exp-sd, ymax=mean_exp+sd), fill="gray") +
  geom_line() +
  geom_point() +
  plot_theme


```
gene_id=="NC_007677.1_WP_011404249.1" is most highly expressed --> putative xanthorhodopsin as per Mongodin et al. 2005 (B) close to lycopene beta-cyclase and a tRNA!

much less expressed:
NC_007677.1_WP_011405484.1 --> inward-directed chlorine pump similar to haloarchaeal rhodopsins (E)
NC_007677.1_WP_011405223.1 --> putative sensory rhodopsins: 
NC_007677.1_WP_011405291.1 --> putative sensory rhodopsins:

These have different labels on the genome browser: it's confusing
"in Halobacterium, two SRs (SRI and SRII) interact to produce a color-sensitive phototactic behavior"
"Until very recently (see below), there has been no information about Salinibacter photobiology and the function of the two genes homologous to SRs still can be inferred only from the genomic context."

NODE_242_length_47958_cov_30.149072_pgaptmp_000939 : the rhodopsin from salinibacter sbsw
BLAST this: aligns to same region of sruber genome as NC_007677.1_WP_011404249.1, putative xanthorhodopsin

#Choosing genes for the heatmap comparing bins to genome

-rhodopsins
NC_007677.1_WP_011404221.1: glycerol kinase
NC_007677.1_WP_011404468.1: superoxide dismutase
NC_007677.1_WP_112903941.1: sufB
NC_007677.1_WP_011404137.1: sufD
NC_007677.1_WP_011405017.1: pstS

S. ruber genome
```{r}
#volcano plot
light_df$diffex <- "padj > 0.05"
light_df$diffex[light_df$padj < 0.05] <- "padj < 0.05"
light_df$diffex[light_df$padj < 0.01] <- "padj < 0.01"
ggplot(light_df, aes(x=log2FoldChange, y=-log10(padj), col=diffex)) + geom_point() +
  scale_color_manual(values=c("blue", "red", "black")) +
  plot_theme

greatest_padj <- head(arrange(light_df,padj), n=35) %>% filter(product != "hypothetical protein")

greatest_lfc_pos <-head(arrange(light_df, desc(log2FoldChange)), n=35) %>% filter(padj <=0.05) %>% filter(product != "hypothetical protein")

greatest_lfc_neg <-head(arrange(light_df, log2FoldChange), n=35) %>% filter(padj <=0.05) %>% filter(product != "hypothetical protein")

genes_of_interest <-light_df %>% filter(gene_id %in% c("NC_007677.1_WP_011404249.1", "NC_007677.1_WP_011405484.1", "NC_007677.1_WP_011405223.1", "NC_007677.1_WP_011405291.1", "NC_007677.1_WP_011404221.1", "NC_007677.1_WP_011404468.1", "NC_007677.1_WP_112903941.1", "NC_007677.1_WP_011404137.1", "NC_007677.1_WP_011405017.1"))

#join dataframes and remove duplicates

genome_heatmap <-rbind(greatest_padj, greatest_lfc_pos, greatest_lfc_neg, genes_of_interest) %>% distinct()

write_csv(genome_heatmap, "sruber_culture_heatmap.csv")

#now find these genes in salinibacter

#test
selected_genes <- salinibacter %>% filter(product %in% genome_heatmap$product | product %in% c("bacteriorhodopsin", "bacteriorhodopsin-like", "glycerol kinase GlpK", "superoxide dismutase", "Fe-S cluster assembly protein SufB", "Fe-S cluster assembly protein SufD", "PstS family phosphate ABC transporter substrate-binding protein"))

test <- selected_genes %>% pivot_wider(., names_from=bin, values_from=product)

write.csv(selected_genes, "salinibacter_sbsw_heatmap.csv")


#download and merge dataframes in excel to compare the selected genes across all genomes

plot <-read_csv("../revised_figures/sruber_heatmap_for_plot.csv")
plot$bin.48 <- as.numeric(plot$bin.48)

plot_long <- plot %>% pivot_longer(!product, names_to="genome", values_to="foldchange")

ggplot(plot_long, aes(x=genome, y=product, fill=foldchange)) +
  geom_tile() +
  scale_fill_viridis() +
  plot_theme

plot <- as.data.frame(plot)
rownames(plot) <-plot$product
plot$product <-NULL
pheatmap(plot, cluster_cols=FALSE)

```

TetR family transcriptional regulator in genome:
See if this has a similar sequence: NODE_6628_length_5362_cov_86.014132_pgaptmp_002059


#Visualizing how transcription level for bins changes across day/night

```{r}

total_counts_sample <- as.data.frame(colSums(normalized_counts[,c(1:36)])) 
total_counts_sample$sample <-rownames(total_counts_sample)


counts_per_bin <- normalized_counts %>% pivot_longer(c(1:36), names_to="sample", values_to="count") %>% group_by(bin, sample) %>% summarize(counts_per_bin=sum(count))

prop_counts <-left_join(counts_per_bin, total_counts_sample, by="sample") %>% mutate(percent_counts_bin=(counts_per_bin/colSums(normalized_counts[,c(1:36)]))*100)

#Join to higher order taxonomic information
#bins <-unique(master_tax$bin)
#taxonomy <-read_csv("../mt_data_v2/gtdbtk_results/gtdbtk_full.csv")
#sbsw_bins <-taxonomy %>% filter(bin %in% bins)
sbsw_bins <-read_csv("sbsw_bins.csv") %>%select(-...1)

prop_counts <-left_join(prop_counts, sbsw_bins[,c(1:8)], by="bin")

prop_counts <- prop_counts %>% mutate(plot=if_else(percent_counts_bin>=1, bin, "Other"))


#visualize this on a plot--transcript counts per sample and their taxonomic assignments

ggplot(prop_counts, aes(x=sample, y=percent_counts_bin, fill=plot)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values=wes_palette_27) +
  plot_theme
#two massively overexpressed genes are driving this trend: "NODE_14322_length_5025_cov_38.148089_pgaptmp_000414", a hypothetical protein, and "NODE_1192_length_23858_cov_130.356636_pgaptmp_000174" (pyruvate dehydrogenase)

#total number of genes ascribed to each sample (not counts/expression level)

genes_in_dataset <- master_tax %>% group_by(bin) %>% summarize(n_genes=length(position))


normalized_counts %>% filter(position=="NODE_14322_length_5025_cov_38.148089_pgaptmp_000414") %>% pivot_longer(c(1:36),names_to="sample", values_to="count") %>% ggplot(., aes(x=sample, y=count)) + geom_point() + plot_theme

normalized_counts %>% filter(position=="NODE_1192_length_23858_cov_130.356636_pgaptmp_000174") %>% pivot_longer(c(1:36),names_to="sample", values_to="count")# %>% #ggplot(., aes(x=sample, y=count)) + geom_point() + plot_theme

#potentially a batch effect?

scale_fill_manual(values=wes_palette_20) 

```

BLAST-checking suspicious genes:

BLAST these genes:
NODE_1943_length_16146_cov_30.276739_pgaptmp_000182 (201005_11_DNA_2_bin.2, Rhodothermia) #16S ribosomal gene
NODE_14322_length_5025_cov_38.148089_pgaptmp_000414 (201007_15_DNA_mat_1_top_bin.31, Halomarina) #no good matches, possibly a ribosome?
NODE_1192_length_23858_cov_130.356636_pgaptmp_000174 (201007_15_DNA_mat_1_top_bin.45, Halobacteriales SW-7-71-33) not a ribosome
NODE_1202_length_16857_cov_174.740507_pgaptmp_000190 (201005_12_DNA_2_bin.1, Halorubrum) not a ribosome
(201007_15_DNA_mat_1_bottom_bin.91) NODE_2703_length_14133_cov_19.928612_pgaptmp_000886: ribosomal RNA gene
(201007_15_DNA_mat_1_top_bin.28) NODE_2592_length_15826_cov_20.447657_pgaptmp_001303: not a ribosome
(201005_12_DNA_2_bin.48) NODE_1402_length_15400_cov_162.203584_pgaptmp_000417: not a ribosome


Spot check important genes of interest on BLAST!!

Filtering out confirmed ribosomal genes: 

```{r}
filtered_counts <-normalized_counts %>% filter(!(position %in% c("NODE_1943_length_16146_cov_30.276739_pgaptmp_000182","NODE_2703_length_14133_cov_19.928612_pgaptmp_000886")))

filtered_total <-as.data.frame(colSums(filtered_counts[,c(1:36)])) 
filtered_total$sample <-rownames(filtered_total)

filtered_counts_bin <- filtered_counts %>% pivot_longer(c(1:36), names_to="sample", values_to="count") %>% group_by(bin, sample) %>% summarize(counts_per_bin=sum(count))

filtered_prop_counts <-left_join(filtered_counts_bin, filtered_total, by="sample") %>% mutate(percent_counts_bin=(counts_per_bin/colSums(filtered_counts[,c(1:36)]))*100)

filtered_prop_counts <-left_join(filtered_prop_counts, sbsw_bins[,c(1:8)], by="bin")

filtered_prop_counts <- filtered_prop_counts %>% mutate(plot=if_else(percent_counts_bin>=0.5, bin, "Other"))


#visualize this on a plot--transcript counts per sample and their taxonomic assignments

ggplot(filtered_prop_counts, aes(x=sample, y=percent_counts_bin, fill=plot)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values=wes_palette_27) +
  plot_theme 



#bubbleplots--dont like these as much


#as a bubble plot rather than a stacked barplot? 
ggplot(filtered_prop_counts, aes(x=sample, y=plot,size=percent_counts_bin)) + 
  geom_point(aes(color=class)) +
  scale_color_manual(values=wes_palette_20) +
  plot_theme

#reorder by domain
filtered_prop_counts$plot_order <-filtered_prop_counts$class
filtered_prop_counts <-filtered_prop_counts %>% mutate(plot_order=if_else(plot=="Other","Other", class))
filtered_prop_counts$plot_order <-as.factor(filtered_prop_counts$plot_order)
levels(filtered_prop_counts$plot_order)

filtered_prop_counts <-filtered_prop_counts %>% left_join(., metadata_norm_v2, by="sample")

#filtered_prop_counts %>% mutate(plot_order=fct_infreq(plot_order)) %>%
  ggplot(filtered_prop_counts, aes(x=sample, y=(plot),size=percent_counts_bin)) + 
  geom_point(aes(color=(plot_order))) +
  scale_color_manual(values=c(wes_palette("Cavalcanti1"), wes_palette("GrandBudapest1"), wes_palette("Darjeeling1"), wes_palette("Moonrise3"), wes_palette("IsleofDogs1"))) +
  plot_theme 
    
```



Plotting the average percentage from all replicates (we lose the batch effect that we are seeing earlier)
Sum the counts across all replicates
Sum the counts per bin across all replicates

```{r}

avg_total <-total_counts_sample %>% left_join(., metadata_norm_v2, by="sample") %>% group_by(time) %>% summarize(total_counts_timepoint=sum(`colSums(normalized_counts[, c(1:36)])`))
```

#Creating Supplemental Table of stats for SBSW bins included in MT analysis

```{r}
bins <-read_csv("sbsw_bins.csv")
all_genomes <-read_csv("../sbsw_bins_summary.csv")

sbsw_bins_stats <-all_genomes %>% filter(genome %in% bins$bin)

write_csv(sbsw_bins_stats, "sbsw_bins_stats.csv")

genes_per_bin <- master_tax %>%group_by(bin) %>% summarize(count=n()) 
write_csv(genes_per_bin, "genes_per_bin.csv")

diffex_genes_bin <-diffex_tax %>% group_by(bin) %>% summarize(count=n())

write_csv(diffex_genes_bin, "diffex_genes_bin.csv")
#add number of genes and diffex genes in dataset

```


